components/select-dropdown/src/useHeadlessSelectHook.ts

```diff
@@ -0,0 +1,128 @@
+import React, { useEffect, useState, useCallback } from "react";
+import { useOutsideClicker, useDebounceEffect, useDidMountEffect } from "hooks";
+
+interface OptionInterface {
+  [key: string]: any;
+}
+
+export interface HeadlessSelectProps {
+  options: {
+    [key: string]: any;
+  }[];
+  valueKey: string;
+  onChange?: (arg: OptionInterface) => void;
+  filterFunction?: (ele: OptionInterface, query: string) => void;
+  clearQuery?: boolean;
+  defaultSelected?: OptionInterface[];
+  defaultOpen?: boolean;
+  defaultQuery?: string;
+  refresh?: any;
+}
+
+export interface HeadlessSelectHook {
+  open: boolean;
+  setOpen: React.Dispatch<React.SetStateAction<boolean>>;
+  query: string;
+  setQuery: React.Dispatch<React.SetStateAction<string>>;
+  selectedValues: OptionInterface[];
+  setSelectedValues: React.Dispatch<React.SetStateAction<OptionInterface[]>>;
+  filteredOptions: OptionInterface[];
+  setFilteredOptions: React.Dispatch<React.SetStateAction<OptionInterface[]>>;
+  addOrRemove: (multiple: boolean, option: OptionInterface) => void;
+  outsideClickRef: React.MutableRefObject<any>;
+}
+
+export default function ({
+  options,
+  valueKey,
+  onChange,
+  filterFunction,
+  clearQuery,
+  defaultSelected = [],
+  defaultOpen = false,
+  defaultQuery = "",
+  refresh = "",
+}: HeadlessSelectProps): HeadlessSelectHook {
+  const [open, setOpen] = useState<boolean>(defaultOpen);
+  const [query, setQuery] = useState<string>(defaultQuery);
+  const [selectedValues, setSelectedValues] =
+    useState<OptionInterface[]>(defaultSelected);
+  const [filteredOptions, setFilteredOptions] = useState<OptionInterface[]>([]);
+
+  const outsideClickRef = useOutsideClicker(() => {
+    setOpen(false);
+  });
+
+  const onTyping = useCallback(() => {
+    if (query) {
+      const val: any = options?.filter((ele: any) => {
+        if (filterFunction) {
+          return filterFunction(ele, query);
+        } else {
+          return (
+            ele.label.toLowerCase().includes(query.toLowerCase()) ||
+            ele.value.toLowerCase().includes(query.toLowerCase())
+          );
+        }
+      });
+      setFilteredOptions([...val]);
+    } else {
+      options && setFilteredOptions([...options]);
+    }
+  }, [query, options]);
+
+  useDebounceEffect(onTyping, 200, [query]);
+
+  useDidMountEffect(() => {
+    setSelectedValues([]);
+    onTyping();
+  }, [refresh]);
+
+  if (clearQuery) {
+    useEffect(() => {
+      !open && setQuery("");
+    }, [open]);
+  }
+
+  const addOrRemove = useCallback(
```

Comment:
@ShreyDhyani This has to be moved into another Hook. This logic can be used by even checkboxes, radio buttons. I think that's how it was in v1.

Reply:
Okay @rjvim. Shall I add it in hooks package and publish those, or create another package for this hook specifically? We have other hooks also which are used across component but are not published yet.

Reply:
@ShreyDhyani For now you can add to the hooks folder. We needn't publish it.